File: LeetCode/C/234-two-pointers-回文链表.cpp
/*
 * @lc app=leetcode.cn id=234 lang=cpp
 *
 * [234] 回文链表
 */
#include "main.h"

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
  public:
    bool isPalindrome(ListNode *head) {
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        if (fast)
            slow = slow->next;
        ListNode *left = head, *right = reverse(slow);
        while (right) {
            if (left->val != right->val)
                return false;
            left = left->next;
            right = right->next;
        }
        return true;
    }

  private:
    ListNode *reverse(ListNode *head) {
        ListNode *p = nullptr, *c = head, *t;
        while (c) {
            t = c->next;
            c->next = p;
            p = c;
            c = t;
        }
        return p;
    }
};
// @lc code=end

File: LeetCode/Go/234-two-pointers-回文链表.go
/*
 * @lc app=leetcode.cn id=234 lang=golang
 *
 * [234] 回文链表
 */
package leetcode

// @lc code=start
func isPalindrome(head *ListNode) bool {
	var slow, fast = head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	if fast != nil {
		slow = slow.Next
	}

	var reverse = func(head *ListNode) *ListNode {
		var p, c, t *ListNode = nil, head, nil
		for c != nil {
			t = c.Next
			c.Next = p
			p = c
			c = t
		}
		return p
	}

	var left, right = head, reverse(slow)
	for right != nil {
		if left.Val != right.Val {
			return false
		}
		left = left.Next
		right = right.Next
	}
	return true
}

// @lc code=end
